File: src/main/java/org/svomz/apps/finances/ports/adapters/persistence/mongo/MongoTransactionRepository.java
package org.svomz.apps.finances.ports.adapters.persistence.mongo;

import static com.mongodb.client.model.Filters.*;

import com.google.common.base.Preconditions;

import com.mongodb.MongoClient;
import com.mongodb.client.MongoCursor;
import com.mongodb.client.MongoDatabase;

import org.bson.Document;
import org.springframework.stereotype.Component;
import org.svomz.apps.finances.domain.model.AccountId;
import org.svomz.apps.finances.domain.model.Expense;
import org.svomz.apps.finances.domain.model.Income;
import org.svomz.apps.finances.domain.model.Tag;
import org.svomz.apps.finances.domain.model.Transaction;
import org.svomz.apps.finances.domain.model.TransactionId;
import org.svomz.apps.finances.domain.model.TransactionRepository;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import javax.inject.Inject;

import static com.mongodb.client.model.Sorts.descending;

/**
 * Created by eric on 18/04/17.
 */
@Component
public class MongoTransactionRepository implements TransactionRepository {

  private final MongoClient mongoClient;
  private final MongoDatabase mongoDatabase;

  @Inject
  public MongoTransactionRepository(MongoClient mongoClient) {
    this.mongoClient = Preconditions.checkNotNull(mongoClient);
    this.mongoDatabase = this.mongoClient.getDatabase("finances");
  }

  @Override
  public void create(Transaction expense) {
    this.mongoDatabase.getCollection("transactions").insertOne(this.map(expense));
  }

  @Override
  public TransactionId nextIdentity() {
    return new TransactionId(UUID.randomUUID().toString());
  }

  @Override
  public List<Transaction> findByAccountId(AccountId accountId, int page, int size) {
    MongoCursor<Document> cursor = this.mongoDatabase.getCollection("transactions")
      .find(eq("accountId", accountId.getId()))
      .sort(descending("date"))
      .limit(size)
      .skip((page - 1)* size)
      .iterator();

    List<Transaction> results = new ArrayList<>();
    while (cursor.hasNext()) {
      results.add(this.map(cursor.next()));
    }

    cursor.close();
    return results;
  }

  @Override
  public List<Transaction> findAllByAccountId(AccountId accountId) {
    MongoCursor<Document> cursor = this.mongoDatabase.getCollection("transactions")
      .find(eq("accountId", accountId.getId())).iterator();

    List<Transaction> results = new ArrayList<>();
    while (cursor.hasNext()) {
      results.add(this.map(cursor.next()));
    }

    cursor.close();
    return results;
  }

  @Override
  public Optional<Transaction> findById(AccountId accountId, TransactionId transactionId) {
    Document document = this.mongoDatabase.getCollection("transactions")
      .find(and(eq("accountId", accountId.getId()), eq("transactionId", transactionId.getId()))).first();

    if (document == null) {
      return Optional.empty();
    }

    return Optional.of(this.map(document));
  }

  @Override
  public void update(Transaction transaction) {
    this.mongoDatabase.getCollection("transactions").updateOne(
      eq("transactionId", transaction.getTransactionId().getId()), new Document("$set", this.map(transaction)));
  }

  @Override
  public void delete(Transaction transaction) {
    this.mongoDatabase.getCollection("transactions")
      .deleteOne(eq("transactionId", transaction.getTransactionId().getId()));
  }

  @Override
  public List<Transaction> findAllByTag(AccountId accountId, Tag tag) {
    MongoCursor<Document> cursor = this.mongoDatabase.getCollection("transactions")
      .find(
        and(
          eq("accountId", accountId.getId()),
          eq("tags", tag.getName())
        )
      ).iterator();

    List<Transaction> results = new ArrayList<>();
    while (cursor.hasNext()) {
      results.add(this.map(cursor.next()));
    }

    cursor.close();
    return results;
  }

  private Transaction map(Document next) {
    double amount = next.getDouble("amount");
    TransactionId transactionId = new TransactionId(next.getString("transactionId"));
    AccountId accountId = new AccountId(next.getString("accountId"));
    BigDecimal value = BigDecimal.valueOf(amount);
    LocalDateTime date =
      LocalDateTime.ofInstant(next.getDate("date").toInstant(), ZoneId.systemDefault());
    String description = next.getString("description");

    Set<Tag> tags = new HashSet<>();
    if (next.get("tags") != null) {
      tags = ((List<String>) next.get("tags")).stream().map(s -> new Tag(s)).collect(Collectors.toSet());
    }

    if (amount < 0) {
      return new Expense(
        transactionId,
        accountId,
        value,
        date,
        description,
        tags
      );
    }
    return new Income(
      transactionId,
      accountId,
      value,
      date,
      description,
      tags
    );
  }

  private Document map(Transaction expense) {
    return new Document("transactionId", expense.getTransactionId().getId())
      .append("accountId", expense.getAccountId().getId())
      .append("amount", expense.value().doubleValue())
      .append("date", Date.from(expense.getDate().atZone(ZoneId.systemDefault()).toInstant()))
      .append("description", expense.getDescription())
      .append("tags", expense.getTags().stream().map(t -> t.getName()).collect(Collectors.toList()));
  }
}

File: web/app/services/port/adapter/secondary/persistence/mongo/MongoTransactionRepository.scala
package services.port.adapter.secondary.persistence.mongo

import java.util.{Date, UUID}

import com.google.inject.Inject
import org.joda.time.DateTime
import org.mongodb.scala.MongoDatabase
import org.mongodb.scala.bson.collection.immutable.Document
import org.mongodb.scala.model.Aggregates._
import org.mongodb.scala.model.Filters._
import org.mongodb.scala.model.Sorts.descending
import org.svomz.apps.finances.core.domain.model._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

import play.api.Logger

class MongoTransactionRepository @Inject()(db: MongoDatabase) extends TransactionRepository {

  private val collection = db.getCollection("transactions")

  override def nextIdentity: Future[TransactionId] = Future { TransactionId(UUID.randomUUID.toString) }

  override def create(transaction: Transaction): Future[Transaction] = {
    collection.insertOne(toDocument(transaction)).toFuture() map { _ => transaction }
  }

  override def update(transaction: Transaction): Future[Transaction] = {
    collection.replaceOne(equal("transactionId", transaction.id.value), toDocument(transaction)).toFuture() map { _ => transaction }
  }

  override def delete(transaction: Transaction): Future[Transaction] = {
    collection.deleteOne(equal("transactionId", transaction.id.value)).toFuture() map { _ => transaction }
  }

  override def fetch(accountId: AccountId, transactionId: TransactionId): Future[Option[Transaction]] = {
    collection.find(equal("transactionId", transactionId.value)).first().toFuture() map { s => s.headOption match {
      case None => None
      case Some(d) => Some(fromDocument(d))
    } }
  }

  override def fetchAll(accountId: AccountId): Future[List[Transaction]] = {
    collection.find(equal("accountId", accountId.value))
      .sort(descending("date"))
      .toFuture() map { s =>
      s map { doc => fromDocument(doc) } toList
    }
  }

  override def fetchAllWithCategory(accountId: AccountId, category: Category, from: Option[Date], to: Option[Date]): Future[List[Transaction]] = {
    collection.find(and(equal("accountId", accountId.value), equal("tags", category.name)))
      .sort(descending("date"))
      .toFuture() map { s =>
      s map { doc => fromDocument(doc) } toList
    }
  }

  override def fetchAllCategories(accountId: AccountId): Future[List[Category]] = {
    collection.aggregate(
      Seq(
        `match`(equal("accountId", accountId.value)),
        unwind("$tags"),
        group("$tags")
      )
    ).toFuture() map { s => {
      s flatMap { i => i map { t => Category(t._2.asString().getValue)} } toList
    } }
  }

  private def toDocument(transaction: Transaction): Document = {
    val tags = transaction.categoryOption match {
      case None => List()
      case Some(c) => List(c.name)
    }

    Document(
      "accountId" -> transaction.accountId.value,
      "transactionId" -> transaction.id.value,
      "amount" -> transaction.amount.toDouble,
      "date" -> transaction.date,
      "description" -> transaction.descriptionOption.getOrElse(""),
      "tags" -> tags
    )
  }

  def fromDocument(doc: Document): Transaction = {
    val accountId: String = doc("accountId").asString().getValue
    val transactionId: String = doc("transactionId").asString().getValue
    val date: Date = new DateTime(doc("date").asDateTime().getValue).toDate
    val amount: BigDecimal = BigDecimal(doc("amount").asDouble().doubleValue())
    val description: String = doc("description").asString().getValue
    val tags = doc("tags").asArray()
    val category: String = if (tags.iterator().hasNext) tags.iterator().next().asString().getValue else ""
    val isDebit: Boolean = amount < 0

    if (isDebit) {
      Debit(
        AccountId(accountId),
        TransactionId(transactionId),
        amount.abs,
        date,
        if (description.isEmpty) None else Some(description),
        if (category.isEmpty) None else Some(Category(category))
      )
    } else {
      Credit(
        AccountId(transactionId),
        TransactionId(transactionId),
        amount.abs,
        date,
        if (description.isEmpty) None else Some(description),
        if (category.isEmpty) None else Some(Category(category))
      )
    }
  }

  override def fetchAllDebitWithCategory(id: AccountId, c: Category, from: Option[Date], to: Option[Date]): Future[List[Transaction]] = ???

  override def fetchAllDebits(id: AccountId, from: Some[Date], to: Some[Date]): Future[List[Transaction]] = ???

  override def fetchAllCredits(id: AccountId, from: Some[Date], to: Some[Date]): Future[List[Transaction]] = ???
}
